### PR TITLE
[EC] Implement generic scalar multiplication

### PR DESCRIPTION
**Pull request by dkostic**
### Issues:
Resolves #ISSUE-NUMBER1
Addresses #ISSUE-NUMBER2

### Description of changes: 
This change implements a generic scalar multiplication function
that can work in four modes of operation which covers the three
functions we need to implement: multiplication of an arbitrary
point, of a base point, and the so-called public multiplication.

### Call-outs:
Point out areas that need special attention or support during the review process. Discuss architecture or design changes.

### Testing:
How is this change tested (unit tests, fuzz tests, etc.)? Are there any testing steps to be verified by the reviewer?

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license and the ISC license.

### CODE REVIEW

File: crypto/fipsmodule/ec/ec_nistp.c [hunk 1]
```diff
@@ -479,22 +498,110 @@ void ec_nistp_scalar_mul(const ec_nistp_meth *ctx,
...
+  ec_nistp_felem_limb *z_res_p = &res_p[ctx->felem_num_limbs * 2];
+
+  // Table of multiples of P = (x_in, y_in, z_in).
+  ec_nistp_felem_limb table_p[SCALAR_MUL_TABLE_MAX_NUM_FELEM_LIMBS];
+  generate_table(ctx, table_p, x_p, y_p, z_p);
+
+  // Table of multiples of the base point G.
+  const ec_nistp_felem_limb *table_g = ctx->scalar_mul_base_table;
+
+  scalar_mul_core(ctx, x_res_g, y_res_g, z_res_g, scalar_g, table_g, base);
```
Comment (torben-hansen) [inline crypto/fipsmodule/ec/ec_nistp.c:602]: non constant modes?

File: crypto/fipsmodule/ec/ec_nistp.c [hunk 2]
```diff
@@ -365,68 +365,77 @@ static inline void select_point_from_table(const ec_nistp_meth *ctx,
...
 //
 // The scalar is recoded (regular-wNAF encoding) into signed digits as explained
 // in |scalar_rwnaf| function. Namely, for a window size |w| we have:
 //     scalar' = s_0 + s_1*2^w + s_2*2^(2*w) + ... + s_{m-1}*2^((m-1)*w),
 // where digits s_i are in [\pm 1, \pm 3, ..., \pm (2^w-1)] and
 // m = ceil(scalar_bit_size / w). Note that for an odd scalar we have that
 // scalar = scalar', while in the case of an even scalar we have that
-// scalar = scalar' - 1.
+// scalar = scalar' - 1. Note that the window size is currently hard-coded for
+// all curves w = SCALAR_MUL_WINDOW_SIZE = 5.
```
[the remark this answers is not in the document]
Comment (dkostic) [reply crypto/fipsmodule/ec/ec_nistp.c:405]: added

File: crypto/fipsmodule/ec/ec_nistp.c [hunk 3]
```diff
@@ -444,6 +453,15 @@ void ec_nistp_scalar_mul(const ec_nistp_meth *ctx,
   ec_nistp_felem_limb *x_tmp = &tmp[0];
   ec_nistp_felem_limb *y_tmp = &tmp[ctx->felem_num_limbs];
   ec_nistp_felem_limb *z_tmp = &tmp[ctx->felem_num_limbs * 2];
```
[the remark this answers is not in the document]
Comment (dkostic) [reply crypto/fipsmodule/ec/ec_nistp.c:458]: few lines above:
`ec_nistp_felem_limb tmp[3 * FELEM_MAX_NUM_OF_LIMBS];`
where `FELEM_MAX_NUM_OF_LIMBS` is defined to ensure no overflow. But I can add an `assert` if you think it's worth it?

File: crypto/fipsmodule/ec/ec_nistp.c (hunk 1 above)
Comment (dkostic) [reply crypto/fipsmodule/ec/ec_nistp.c]: right :)

File: crypto/fipsmodule/ec/ec_nistp.c [hunk 4]
```diff
@@ -365,68 +368,77 @@ static inline void select_point_from_table(const ec_nistp_meth *ctx,
...
-  ec_nistp_felem_limb table[SCALAR_MUL_TABLE_MAX_NUM_FELEM_LIMBS];
-  generate_table(ctx, table, x_in, y_in, z_in);
+static inline void scalar_mul_core(const ec_nistp_meth *ctx,
+                                   ec_nistp_felem_limb *x_out,
+                                   ec_nistp_felem_limb *y_out,
+                                   ec_nistp_felem_limb *z_out,
+                                   const EC_SCALAR *scalar,
+                                   const ec_nistp_felem_limb *table,
+                                   enum scalar_mul_core_mode mode) {
+  // Mode based invariants.
```
Comment (torben-hansen) [inline crypto/fipsmodule/ec/ec_nistp.c:437]: Just something to ponder later. Is something else using `scalar_mul_core` apart from the three function further below? If not, while these modes attempts to abstract some semantics, it's almost too much abstraction. If the set of functions that invoke, and will ever invoke `scalar_mul_core`, one could have this as three arguments.

It also appears there is some implicit relationship between the table argument and the mode argument. Although, any mistakes mis-configuring that e.g. use `mode != base` for a full table, will likely cause the unit tests to have severe hiccups.

File: crypto/fipsmodule/ec/ec_nistp.c (hunk 3 above)
[the remark this answers is not in the document]
Comment (torben-hansen) [reply crypto/fipsmodule/ec/ec_nistp.c:458]: Actually, you already have these below
```
  assert(SCALAR_MUL_TABLE_MAX_NUM_FELEM_LIMBS >=
         SCALAR_MUL_TABLE_NUM_POINTS * ctx->felem_num_limbs * 3);
```
So, that should be sufficient, right?

**Review by torben-hansen — APPROVED** [review]

**Review by nebeid — COMMENTED** [review]
Closed, but just submitting my pending changes.

**Pull request closed without merging**